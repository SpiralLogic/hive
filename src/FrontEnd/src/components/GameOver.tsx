import { FunctionComponent, Fragment } from 'preact';
import '../css/share.css';

const GameOver: FunctionComponent<{ outcome: string }> = ({ outcome }) => {
  return (
    <Fragment>
      <p>{outcome}</p>
      <button autofocus title="New Game" onClick={() => window.location.assign(`/`)}>
        Close
      </button>
    </Fragment>
  );
};
GameOver.displayName = 'GameOver';
export default GameOver;
