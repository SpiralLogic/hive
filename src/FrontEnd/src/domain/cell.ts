import { HexCoordinates } from './hex-coordinates';
import { Tile } from './tile';

export type Cell = {
  coords: HexCoordinates;
  tiles: Tile[];
};
