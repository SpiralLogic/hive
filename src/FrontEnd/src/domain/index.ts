export * from './cell';
export * from './game-state';
export * from './move';
export * from './tile';
export * from './player';
export * from './cell';
export * from './Row';
export * from './hex-coordinates';
