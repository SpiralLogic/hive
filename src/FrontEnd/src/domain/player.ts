import { Tile } from './tile';

export type Player = {
  id: PlayerId;
  name: PlayerName;
  tiles: Tile[];
};

export type PlayerId = number;
export type PlayerName = string;
