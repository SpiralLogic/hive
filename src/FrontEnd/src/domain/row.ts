import { Cell } from './cell';

export type Row = { id: number; hidden: boolean; cells: Array<Cell & { hidden?: boolean }> };
