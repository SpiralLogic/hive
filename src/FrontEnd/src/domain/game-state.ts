import { Cell } from './cell';
import { Player } from './player';

export type GameState = {
  cells: Cell[];
  players: Player[];
  gameId: GameId;
  gameStatus: GameStatus
};

export type GameId = string;
export type GameStatus =
  | 'NewGame'
  | 'MoveSuccess'
  | 'AiWin'
  | 'Player0Win'
  | 'Player1Win'
  | 'MoveSuccessNextPlayerSkipped'
  | 'MoveInvalid'
  | 'GameOver' | 'Draw';
