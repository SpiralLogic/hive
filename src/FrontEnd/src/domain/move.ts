import {GameId} from "./game-state";
import {HexCoordinates} from './hex-coordinates';
import {TileId} from './tile';

export type Move = {
    tileId: TileId;
    coords: HexCoordinates;
};

export type MoveTile = (gameId: GameId, move: Move) => void;
