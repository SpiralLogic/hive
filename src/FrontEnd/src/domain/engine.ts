import { GameId, GameState } from './game-state';
import { HubConnectionState } from '@microsoft/signalr';
import { Move } from './move';
import { Tile } from './tile';

export type GameStateUpdateHandler = (gameState: GameState) => void;
export type OpponentSelectionHandler = (type: 'select' | 'deselect', tile: Tile) => void;
export type GameConnection = {
  getConnectionState: () => HubConnectionState;
  closeConnection: () => Promise<void>;
  sendSelection: OpponentSelectionHandler;
};
export type GameHandlers = {
  updateGameState: GameStateUpdateHandler;
  opponentSelection?: OpponentSelectionHandler;
};
export type HexEngine = {
  connectGame: (gameId: GameId, handlers: GameHandlers) => GameConnection;
  getNewGame: () => Promise<GameState>;
  moveTile: (gameId: GameId, move: Move) => Promise<GameState>;
  getExistingGame: (gameId: GameId) => Promise<GameState>;
};
