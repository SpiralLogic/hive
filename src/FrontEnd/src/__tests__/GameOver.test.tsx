import { render, screen } from '@testing-library/preact';
import userEvent from '@testing-library/user-event';
import { h } from 'preact';
import GameOver from '../components/GameOver';
import { GameStatus } from '../domain';
import { mockLocation, renderElement } from './test-helpers';

describe('gameOver snapshot tests', () => {
  test('close calls base url', () => {
    const restore = mockLocation({ assign: jest.fn() });
    render(<GameOver outcome="Draw" />);

    userEvent.click(screen.getByRole('button'));
    expect(window.location.assign).toBeCalledWith('/');
    restore();
  });

  const outcomes: GameStatus[] = [
    'NewGame',
    'MoveSuccess',
    'AiWin',
    'Player0Win',
    'Player1Win',
    'MoveSuccessNextPlayerSkipped',
    'MoveInvalid',
    'GameOver',
    'Draw',
  ];

  outcomes.forEach((outcome) => {
    it(`${outcome} snapshot player 0`, () => {
      expect(renderElement(<GameOver outcome={outcome} />)).toMatchSnapshot();
    });

    it(`${outcome} snapshot player 1`, () => {
      expect(renderElement(<GameOver outcome={outcome} />)).toMatchSnapshot();
    });
  });
});
