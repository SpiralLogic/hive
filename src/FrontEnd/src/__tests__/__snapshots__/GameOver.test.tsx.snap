// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gameOver snapshot tests AiWin snapshot player 0 1`] = `
<p>
  AiWin
</p>
`;

exports[`gameOver snapshot tests AiWin snapshot player 1 1`] = `
<p>
  AiWin
</p>
`;

exports[`gameOver snapshot tests Draw snapshot player 0 1`] = `
<p>
  Draw
</p>
`;

exports[`gameOver snapshot tests Draw snapshot player 1 1`] = `
<p>
  Draw
</p>
`;

exports[`gameOver snapshot tests GameOver snapshot player 0 1`] = `
<p>
  GameOver
</p>
`;

exports[`gameOver snapshot tests GameOver snapshot player 1 1`] = `
<p>
  GameOver
</p>
`;

exports[`gameOver snapshot tests MoveInvalid snapshot player 0 1`] = `
<p>
  MoveInvalid
</p>
`;

exports[`gameOver snapshot tests MoveInvalid snapshot player 1 1`] = `
<p>
  MoveInvalid
</p>
`;

exports[`gameOver snapshot tests MoveSuccess snapshot player 0 1`] = `
<p>
  MoveSuccess
</p>
`;

exports[`gameOver snapshot tests MoveSuccess snapshot player 1 1`] = `
<p>
  MoveSuccess
</p>
`;

exports[`gameOver snapshot tests MoveSuccessNextPlayerSkipped snapshot player 0 1`] = `
<p>
  MoveSuccessNextPlayerSkipped
</p>
`;

exports[`gameOver snapshot tests MoveSuccessNextPlayerSkipped snapshot player 1 1`] = `
<p>
  MoveSuccessNextPlayerSkipped
</p>
`;

exports[`gameOver snapshot tests NewGame snapshot player 0 1`] = `
<p>
  NewGame
</p>
`;

exports[`gameOver snapshot tests NewGame snapshot player 1 1`] = `
<p>
  NewGame
</p>
`;

exports[`gameOver snapshot tests Player0Win snapshot player 0 1`] = `
<p>
  Player0Win
</p>
`;

exports[`gameOver snapshot tests Player0Win snapshot player 1 1`] = `
<p>
  Player0Win
</p>
`;

exports[`gameOver snapshot tests Player1Win snapshot player 0 1`] = `
<p>
  Player1Win
</p>
`;

exports[`gameOver snapshot tests Player1Win snapshot player 1 1`] = `
<p>
  Player1Win
</p>
`;
